File: java/src/utils/TreeNode.java
package utils;

/**
 * Created by 36191 on 2017/12/5
 * 二叉树
 */

public class TreeNode {
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode(int x) { val = x; }

    /**
     * 前序遍历输出
     * @param root
     */
    public static void output(TreeNode root) {
        if (root == null) {
            System.out.print("null, ");
        } else {
            System.out.print(root.val + ", ");
            output(root.left);
            output(root.right);
        }
    }
}

File: golang/treenode.go
package main

import (
	"bytes"
	"fmt"
)

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}


func (t *TreeNode) Equals(tt *TreeNode) bool {
	if t == nil && tt == nil {
		return true
	}
	if t == nil || tt == nil || t.Val != tt.Val {
		return false
	}

	return t.Left.Equals(tt.Left) && t.Right.Equals(tt.Right)
}

func treeNodeToString(root *TreeNode, buffer bytes.Buffer) {
	if root == nil {
		if buffer.Len() == 0 {
			buffer.WriteString("nil")
		} else {
			buffer.WriteString(",nil")
		}
		return
	}

	if buffer.Len() == 0 {
		buffer.WriteString(fmt.Sprintf("%d", root.Val))
	} else {
		buffer.WriteString(fmt.Sprintf(",%d", root.Val))
	}

	treeNodeToString(root.Left, buffer)
	treeNodeToString(root.Right, buffer)
}


func (t *TreeNode) ToString() string {
	var buffer bytes.Buffer
	treeNodeToString(t, buffer)
	return buffer.String()
}
